contracts/contracts/GrantRegistry.sol

```diff
@@ -1,135 +1,124 @@
+   */
+  function updateGrantPayee(uint96 _id, address _payee) external {
+    Grant storage grant = grants[_id];
+    require(msg.sender == grant.owner, "Not authorized");
+    grant.payee = _payee;
+    emit GrantUpdated(grant.id, grant.owner, grant.payee, grant.metaPtr);
   }
 
-  function getGrantCount() public view returns (uint256) {
-    return grantCount;
+  /**
+   * @notice Update the payee of a grant
+   * @param _id ID of grant to update
+   * @param _metaPtr New URL that points to grant metadata
+   */
+  function updateGrantMetaPtr(uint96 _id, string calldata _metaPtr) external {
+    Grant storage grant = grants[_id];
+    require(msg.sender == grant.owner, "Not authorized");
+    grant.metaPtr = _metaPtr;
+    emit GrantUpdated(grant.id, grant.owner, grant.payee, grant.metaPtr);
   }
 
-  function getAllGrantIds() public view returns (bytes32[] memory) {
-    bytes32[] memory ret = new bytes32[](grantCount);
-    for (uint256 i = 0; i < grantCount; i++) {
-      ret[i] = grantIds[i];
-    }
-    return grantIds;
+  /**
+   * @notice Update multiple fields of a grant at once
+   * @dev To leave a field unchanged, you must pass in the same value as the current value
+   * @param _id ID of grant to update
+   * @param _owner New owner address
+   * @param _payee New payee address
+   * @param _metaPtr New URL that points to grant metadata
+   */
+  function updateGrant(
+    uint96 _id,
+    address _owner,
+    address _payee,
+    string calldata _metaPtr
+  ) external {
+    Grant storage grant = grants[_id];
+    require(msg.sender == grant.owner, "Not authorized");
+    grant.owner = _owner;
+    grant.payee = _payee;
+    grant.metaPtr = _metaPtr;
+    emit GrantUpdated(grant.id, grant.owner, grant.payee, grant.metaPtr);
   }
 
+  // --- View functions ---
+  /**
+   * @notice Returns an array of all grants and their on-chain data
+   * @dev May run out of gas for large values `grantCount`, depending on the node's RpcGasLimit. In these cases,
+   * `getGrants` can be used to fetch a subset of grants and aggregate the results of various calls off-chain
+   */
   function getAllGrants() public view returns (Grant[] memory) {
-    Grant[] memory ret = new Grant[](grantCount);
-    bytes32 grantId;
-    for (uint256 i = 0; i < grantCount; i++) {
-      grantId = grantIds[i];
-      ret[i] = grants[grantId];
-    }
-    return ret;
+    return getGrants(0, grantCount);
   }
 
-  /*
-  function getAllGrants() public view returns (myGrant[] memory) {
-    grantCount = getGrantCount();
-    myGrant[] memory ret = new myGrant[](grantCount);
-    for (uint i = 0; i < grantCount; i++) {
-      ret[i] = grants[i];
+  /**
+   * @notice Returns a range of grants and their on-chain data
+   * @param _startId Grant ID of first grant to return, inclusive, i.e. this grant ID is included in return data
+   * @param _endId Grant ID of last grant to return, exclusive, i.e. this grant ID is NOT included in return data
```

Comment:
why exclude the grant with id `_endId` in the returned grant array? i can see that it saves some typing in that we can use `grantCount` instead of `grantCount-1`, but as a con it adds complexity to consumers of this method i.e. now i have to remember the upper boundary is exclusive.

Reply:
This format of "inclusive start index, exclusive end index" seems to be the standard way to slice arrays, which I why I did it. For example, JavaScript's `slice`, Python's list/array indexing, and Solidity's array slicing all work this way

Reply:
Yeah, I think this convention stems from `0` indexing— since the `length` of an array is always 1 greater than the largest index position, the second index is conventionally non-inclusive, so if you want the whole array you ask for 0 to length. I also find it confusing @wildmolasses but let's keep it as-is since this is the convention used by the languages most devs interacting with this contract will be familiar with.

Reply:
Go is like this as well! But R is not! 🤷‍♂️  OK, sounds good!